### PR TITLE
Unified memory manager and other minor fixes

### LINKED ISSUE #1214: Create a unified memory manager class / framework for all backends
This way we do not need to replicate the changes done in all three backends.

### PR DESCRIPTION
**Pull request by pavanky**
- src/backend/MemoryManager.{hpp,cpp} unify memory managers from all backends.
- Moved dispatch.{hpp,cpp} moved to src/backend
- Renamed mem{Pop/Push} --> mem{Lock/Unlock}
- Fixed issue with sort in CUDA device manager

- Fixes #1214

### CODE REVIEW

**Comment by pavanky** [comment]
Output of af::info **before** the CUDA device sort change:

```
ArrayFire v3.3.0 (CUDA, 64-bit Linux, build 09bc060)
Platform: CUDA Toolkit 7.5, Driver: 358.16
[0] Tesla K20c, 4800 MB, CUDA Compute 3.5
-1- Tesla K20c, 5120 MB, CUDA Compute 3.5
-2- Quadro 6000, 6144 MB, CUDA Compute 2.0
```

Output of af::info **after** the CUDA device sort change:

```
ArrayFire v3.3.0 (CUDA, 64-bit Linux, build 73b7cac)
Platform: CUDA Toolkit 7.5, Driver: 358.16
[0] Tesla K20c, 5120 MB, CUDA Compute 3.5
-1- Tesla K20c, 4800 MB, CUDA Compute 3.5
-2- Quadro 6000, 6144 MB, CUDA Compute 2.0
```

Note the K20c with 5120MB is now at top. Before the fix, both of them had the same flop count so the sort was just picking the first device it encountered without considering the memory available.

**Comment by shehzan10** [comment]
build arrayfire ci

File: src/backend/MemoryManager.cpp [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+    std::string env_var = getEnvVar("AF_MEM_DEBUG");
+    if (!env_var.empty()) {
+        this->debug_mode = env_var[0] != '0';
+    }
+    if (this->debug_mode) mem_step_size = 1;
+}
+
+void MemoryManager::garbageCollect()
+{
+    if (this->debug_mode) return;
```
Comment (umar456) [inline src/backend/MemoryManager.cpp]: This should require a lock

File: src/backend/MemoryManager.cpp [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+                          true,
+                          100 }; //This number is not relevant
+
+        current.map[(void *)ptr] = info;
+    }
+}
+
+void MemoryManager::userUnlock(const void *ptr)
+{
+    memory_info& current = this->getCurrentMemoryInfo();
```
Comment (umar456) [inline src/backend/MemoryManager.cpp]: This should also be protected by the lock.

File: src/backend/MemoryManager.cpp [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+        current.lock_bytes += alloc_bytes;
+        current.lock_buffers++;
+        current.total_bytes += alloc_bytes;
+    }
+    return ptr;
+}
+
+void MemoryManager::userLock(const void *ptr)
+{
+    memory_info& current = this->getCurrentMemoryInfo();
```
Comment (umar456) [inline src/backend/MemoryManager.cpp]: Should be protected by the lock

File: src/backend/MemoryManager.cpp [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+void *MemoryManager::alloc(const size_t bytes)
+{
+    memory_info& current = this->getCurrentMemoryInfo();
+
+    void *ptr = NULL;
+    size_t alloc_bytes = this->debug_mode ? bytes : (divup(bytes, mem_step_size) * mem_step_size);
+
+    if (bytes > 0) {
+
+        lock_guard_t lock(this->memory_mutex);
```
Comment (umar456) [inline src/backend/MemoryManager.cpp]: probably should move this up to the beginning of the function

**Comment by shehzan10** [comment]
build arrayfire ci

**Comment by shehzan10** [comment]
build arrayfire ci

**Pull request merged**